### ISSUE TITLE
npm3 installs incorrect dependencies when updating using symlinks

### ISSUE THREAD

**Issue by mlucool** [npm/npm#14029]
#### I'm opening this issue because:
- [ ] npm is crashing.
- [X] npm is producing an incorrect install.
- [ ] npm is doing something I don't understand.
- [ ] Other (_see below for feature requests_):
#### What's going wrong?

When using symlinks and you change a the version of a dependency, npm will rightfully complain it will not update it, yet still installs new dependencies to it leaving the tree in a bad state.

I.e. If I am dependent on v1 via symlink and npm install foo@v2 it will tell me that it cannot update but will then install all of v2's missing dependencies to v1 without updating v1.

It should never under any circumstance change the contents of the symlink, as the warning implies was the intention.
#### How can the CLI team reproduce the problem?

Create the following:
A directory called foo with the following package.json:

```
mkdir foo bar
echo '{
  "name": "foo",
  "version": "1.0.0",
  "dependencies": {
    "babel": "^6.5.2",
    "babel-core": "^6.14.0",
    "jsdoc-babel": "^0.2.1"
  }
}
' > foo/package.json
echo '{
   "name": "bar",
   "version": "1.0.0",
   "dependencies": {
     "jsdoc-babel": "^0.1.0"
   }
 }
' > bar/package.json
cd foo
npm install
cd ../bar
mkdir node_modules
cd node_modules
ln -s ../../foo/node_modules/* .
cd ..
npm install
```

The output of this install includes:

```
npm WARN update-linked node_modules/jsdoc-babel needs updating to 0.1.0 from 0.2.1 but we can't, as it's a symlink
bar@1.0.0 /my/path/to/bar/bar
\u2514\u2500\u252c jsdoc-babel@0.1.0
  \u251c\u2500\u252c babel-core@5.8.38 
  \u2502 \u251c\u2500\u2500 babel-plugin-constant-folding@1.0.1 
  \u2502 \u251c\u2500\u2500 babel-plugin-dead-code-elimination@1.0.2 
```

Which shows it is modifying the contents of the symlink!
As a note, the above is also wrong, jsdoc-babel is still v0.2.1:

```
grep version node_modules/jsdoc-babel/package.json
  "version": "0.2.1"
```

### supporting information:

Linux
- `npm -v` prints:  3.10.3
- `node -v` prints: v6.5.0
- `npm config get registry` prints:
- Windows, OS X, or Linux?: Linux
- Network issues:
  - Geographic location where npm was run:
  - [X] I use a proxy to connect to the npm registry.
  - [ ] I use a proxy to connect to the web.
  - [ ] I use a proxy when downloading Git repos.
  - [ ] I access the npm registry via a VPN
  - [ ] I don't use a proxy, but have limited or unreliable internet access.
- Container:
  - [ ] I develop using Vagrant on Windows.
  - [ ] I develop using Vagrant on OS X or Linux.
  - [ ] I develop / deploy using Docker.
  - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by mlucool** [comment]
Hi,

Any update on verifying this?

**Comment by mlucool** [comment]
@iarna would you mind taking a look? It seems related to: https://github.com/npm/npm/commit/bfe37e01a1e470eb2f076d1f9e6199fdd32ef43f

**Comment by iarna** [comment]
The bug here is npm inappropriately installing dependencies inside a symlinked module (when what it _actually_ wanted to do was remove the symlink and install the module fresh.)

Given:

```
depA@1
  dependency: depB@1
debA@2
  dependency: depB@2
```
And a starting testmodule with:

```
dependency: depA@2
 node_modules:
   depA@1 -> symlink
   depB@1
```

Running `npm install` ends up installing depB@2 inside the destination of depA@1, then complaining that it won't update depA@1 (instead of replacing it with depA@2).

My preferred fix for this will be #9519.

**Comment by mlucool** [comment]
Thanks for looking into this!

**Comment by mlucool** [comment]
This appears to be fixed in this example with 5.0.4. @iarna do you believe this to be generically fixed for all cases now? I ask as #9519 is still open.
